[#5201] feat(client-python): Implement expressions in python client

SophieTech88 opened the pull request:
### What changes were proposed in this pull request?
Implement expression from java, including:
- Expression.java
- FunctionExpression.java
- NamedReference.java
- UnparsedExpression.java
- literals/

convert to python client, and add unit test for each class.

### Why are the changes needed?

We need to support the expressions in python client

Fix: #5201

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Need to pass all unit tests.

tengqm commented on clients/client-python/gravitino/api/expressions/named_reference.py:
```diff
@@ -0,0 +1,76 @@
+
+    def __init__(self, field_name: List[str]):
+        super().__init__()
+        self._field_name = field_name
```
rename `_field_name` to `_field_names` ?
Looks like this field is an array.

SophieTech88 replied:
Yeah. Just updated it.

tengqm commented on clients/client-python/gravitino/api/expressions/named_reference.py:
```diff
@@ -0,0 +1,76 @@
+        super().__init__()
+        self._field_name = field_name
+
+    def field_name(self) -> List[str]:
```
```suggestion
    def field_names(self) -> List[str]:
```

SophieTech88 replied:
It is a good suggestion. Just updated the `field_name` to `field_names`.

tengqm replied:
Not resolved?

tengqm commented on clients/client-python/gravitino/api/expressions/named_reference.py:
```diff
@@ -0,0 +1,76 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Why do we call this a named reference?
Is there a case where a reference has no name?
Do we have to differentiate these two types of references?

SophieTech88 replied:
@xunliu  Do you have any idea about the name for this feature?

xunliu replied:
hi @tengqm 
because we need to implement all classes in the Python client.
- https://github.com/apache/gravitino/blob/main/api/src/main/java/org/apache/gravitino/rel/expressions/NamedReference.java

xunliu replied:
@mchades Do you have any suggestions?

tengqm replied:
Really appreciate this clarification. Thanks.

tengqm commented on clients/client-python/gravitino/api/expressions/function_expression.py:
```diff
@@ -0,0 +1,90 @@
+                self._function_name == other._function_name
+                and self._arguments == other._arguments
+            )
+        return False
```
Gut feeling is that you may want to leave a TODO here ...

SophieTech88 replied:
Thanks for your comment. Updated the TODO comment. Does that work for you?

xunliu commented:
hi @SophieTech88 I will help you improve this PR.
Please review my commit.

tengqm commented on clients/client-python/gravitino/api/expressions/Literals/literals.py:
```diff
@@ -0,0 +1,138 @@
+
+    def __init__(
+        self,
+        value: Union[int, float, str, datetime, time, date, bool, Decimal, None],
```
Why do we have a `Decimal` here?

xunliu replied:
Because we need to support `Decimal` type here.
- https://github.com/apache/gravitino/blob/main/api/src/main/java/org/apache/gravitino/rel/expressions/literals/Literals.java#L134

tengqm replied:
Got it.

tengqm reviewed (commented):
I'd suggest we split this into a few smaller PRs.

SophieTech88 commented on clients/client-python/gravitino/api/expressions/Literals/literal.py:
```diff
@@ -0,0 +1,39 @@
+        pass
+
+    @abstractmethod
+    def data_type(self):
```
Agree. Just updated the code to raise NotImplementedError() for those 2 functions.

mchades reviewed (commented):
> I'd suggest we split this into a few smaller PRs.

Agree with @tengqm, the current PR is too large to review. 

As this PR is focused on expressions, I suggest moving `distributions`, `sorts`, and `transforms` to separate PRs. This will make it easier for us to review this PR. WDYT? @SophieTech88

SophieTech88 commented:
> > I'd suggest we split this into a few smaller PRs.
> 
> Agree with @tengqm, the current PR is too large to review.
> 
> As this PR is focused on expressions, I suggest moving `distributions`, `sorts`, and `transforms` to separate PRs. This will make it easier for us to review this PR. WDYT? @SophieTech88

Of course. Just update the PR, removed  `distributions`, `sorts`, and `transforms` .

mchades commented on clients/client-python/gravitino/api/expressions/literals/literals.py:
```diff
@@ -0,0 +1,138 @@
+
+    _value: Union[int, float, str, datetime, time, date, bool, Decimal, None]
+    _data_type: (
+        str  # TODO: Need implement `api/src/main/java/org/apache/gravitino/rel/types`
```
you already implemented types `clients/client-python/gravitino/api/types.py` and type `clients/client-python/gravitino/api/type.py`

SophieTech88 replied:
Oh. Yeah. Do I need to import  and  list all types in this _data_type?

SophieTech88 replied:
I just update the _data_type in the new commit

mchades commented on clients/client-python/gravitino/api/expressions/literals/literals.py:
```diff
@@ -0,0 +1,149 @@
+    """Creates a literal with the given type value."""
+
+    _value: Union[int, float, Decimal, str, date, time, datetime, bool, None]
+    _data_type: Union[
```
seems should be `_data_type: Type`? @xunliu cc

xunliu replied:
I improve this part codes
@SophieTech88 @noidname01  Please help me review, thanks.

SophieTech88 replied:
The changes for the types things, LGTM. Many thanks!

noidname01 commented:
Overall LGTM!

noidname01 commented on clients/client-python/gravitino/api/expressions/expression.py:
```diff
@@ -0,0 +1,51 @@
+class Expression(ABC):
+    """Base class of the public logical expression API."""
+
+    EMPTY_EXPRESSION: List[Expression] = []
```
This might cause unexpected behaviour if it will be modified at any moment (like append, delete, e.g.), because this list would share between classes, any modification would affect where it exists once and for all at the same time. Unless we're sure that the scenario above would not happen, I think it's better to use immutable object here or some other methods.

xunliu replied:
hi @noidname01 
The `EMPTY_EXPRESSION` value does not fill in any data.
So, I think we didn't need to modify this code, what's do you think?

noidname01 replied:
Sure, then overall LGTM!

mchades commented on clients/client-python/gravitino/api/types/decimal.py:
this is unnecessary since https://github.com/apache/gravitino/pull/5354#discussion_r1828647646

xunliu replied:
@mchades I removed `Decimal` type. Please help me review again, Thanks.

mchades reviewed (approved):
Overall LGTM. Do you want to have another look? @tengqm

jerryshao commented:
LGTM.

xunliu commented:
hi @tengqm 
Do you want to have another look?

tengqm commented on clients/client-python/gravitino/api/expressions/function_expression.py:
```diff
@@ -0,0 +1,95 @@
+    @abstractmethod
+    def function_name(self) -> str:
+        """Returns the function name."""
+        pass
```
As noted previously, please reconsider all these `pass` cases ...
do we really want to silently succeed here?

xunliu replied:
Yes, You are right, I will fix this problem.

tengqm commented on clients/client-python/gravitino/api/expressions/literals/literals.py:
```diff
@@ -0,0 +1,143 @@
+    _value: T
+    _data_type: Type
+
+    def __init__(
+        self,
+        value: T,
+        data_type: Type,
+    ):
```
```suggestion
    def __init__(self, value: T, data_type: Type):
```

tengqm commented on clients/client-python/gravitino/api/types/types.py:
```diff
@@ -123,7 +123,7 @@ def get(cls) -> "ShortType":
 
         @classmethod
-        def unsigned(cls):
+        def unsigned(cls) -> "ShortType":
```
Can you help share with me what this syntax is?
I mean ... why the return type is quoted?

SophieTech88 replied:
```
    class ShortType(IntegralType):
        _instance: "ShortType" = None
        _unsigned_instance: "ShortType" = None
   ```
The type annotation ShortType is quoted because `_instance` is an attribute of the class that references ShortType, and the class itself isn't fully defined at the time this attribute is declared.

The return type annotations (-> "ShortType") are also quoted for the same reason: `ShortType` isn't fully defined when these methods are declared.

SophieTech88 replied:
If we remove the quote, there is a Pylance errors, and it won't pass the pylint check.

tengqm commented on clients/client-python/gravitino/api/expressions/expression.py:
```diff
@@ -0,0 +1,51 @@
+
+from __future__ import annotations
+from abc import ABC, abstractmethod
+from typing import List, Set, TYPE_CHECKING
```
I'm new to this `typing` thing.
But according to Python docs, these alias are all deprecated in Python 3.9.
cf. https://docs.python.org/3/library/typing.html#deprecated-aliases

It might not be a good idea to use them, IIUC.

SophieTech88 replied:
Yeah. List, Set are deprecated from typing in Python 3.9. We can use list and set directly.

xunliu commented on clients/client-python/gravitino/api/expressions/function_expression.py:
```diff
@@ -0,0 +1,95 @@
+    A concrete implementation of the FunctionExpression interface.
+    """
+
+    _function_name: str
```
> for a member that is well scoped, I don't think we need a sophisticated name.
> I'd suggest we use _name as the property name rather than _function_name.

and 

> similarly, given an instance impl of class FuncExpressionImpl,
> the following calls are still pretty clear without the complex names:

hi @tengqm 
Will want to keep Python code consistent with Java code, https://github.com/apache/gravitino/blob/main/api/src/main/java/org/apache/gravitino/rel/expressions/FunctionExpression.java#L66

SophieTech88 commented:
The Integration Test failure is related to the List typing things in literal.py file.

xunliu commented:
hi @tengqm We are improving PR base on your comments,
Please help me review it again, thanks!

tengqm commented on clients/client-python/gravitino/api/expressions/literals/literal.py:
```diff
@@ -0,0 +1,43 @@
+# under the License.
+
+from abc import abstractmethod
+from typing import List, TypeVar, Generic
```
You may want to kill 'List' here.

tengqm commented on clients/client-python/gravitino/api/expressions/literals/literals.py:
```diff
@@ -0,0 +1,137 @@
+        return f"LiteralImpl(value={self._value}, data_type={self._data_type})"
+
+
+class Literals:
```
Since we already have a 'Literal' class in another file,
defining a new class called 'Literals' may confuse other developers/users.

tengqm replied:
Further more, it doesn't make a lot senses to me to define a class as a wrapper
for so many class methods.
Python is not a pure object-oriented language, after all.

xunliu replied:
hi @tengqm I know what you mean.
We also refer to Iceberg's Python client code style.
https://github.com/apache/iceberg-python/blob/main/pyiceberg/table/name_mapping.py
At present, there are not enough experienced Python engineers in the community, which is the result of the best efforts of the community partners.
We keep the python client code consistent with the java code to ensure that the functionality is consistent.
So I think we didn't modify these codes.

xunliu commented:
When I use `list` to replace `List`.
The `TestLiterals` throw these errors.
```
The error TypeError: 'type' object is not subscriptable occurs because the list type is not subscriptable in Python versions earlier than 3.9. To fix this, you should use List from the typing module instead of list.
```

@tengqm @noidname01  Do you have any suggestions to fix this problem? I rollback this change??

xunliu commented:
hi @tengqm 
Gravitino Python client went to support from 3.7 to 3.10, So maybe we must use `typing.List` type to support low Python version.

tengqm commented:
> hi @tengqm Gravitino Python client went to support from 3.7 to 3.10, So maybe we must use `typing.List` type to support low Python version.

In that case, we have to stick to the old way.
3.7-3.10 is a large range anyway.

xunliu commented:
hi @tengqm :-)
I revert `remove List` commit, Please help me review it again, Thanks!

mchades reviewed (approved):
I am merging this into the main branch. Thank you all for your efforts on this PR!
@SophieTech88 @tengqm @noidname01  @xunliu
